doc/src/contributing/new-contributors-guide/dev-setup.md

```diff
@@ -0,0 +1,228 @@
+
+Read instructions in '~/.git-completion.sh'
+
+Note that if you install git from the package manager in many Linux distros, this file is already installed for you.  You can check if it is installed by seeing if tab completion works on git commands (try, e.g., `git commi<TAB>`, or `git log --st<TAB>`). You can also check if the PS1 commands work by doing something like:
+
+```
+PS1='\W $(__git_ps1 "%s")\$ '
+```
+
+And your command prompt should change to something like:
+
+```
+sympy master$
+```
+
+Note, it is important to define your PS1 using single quotes ('), not double quotes ("), or else bash will not update the branch name.
+
+## Setup GitHub
+
+Next you will need to setup your GitHub account. Note that all the steps here
+only need to be done once. If you already have a GitHub account and have setup
+SSH keys, even if it was for a different project than SymPy, you do not need
+to do them again.
+
+(dev-setup-create-github-account)=
+### Create a GitHub Account
+
+A [GitHub](https://github.com) account is required to contribute to SymPy. If
+you have not one yet then sign up at <https://github.com/signup/free>. Your
+GitHub account is your presence in the open source world, so we recommend
+choosing a professional username.
+
+### Setup SSH Keys
+
+To establish a secure connection between your computer and GitHub see detailed
+instructions in <https://help.github.com/en/articles/set-up-git> or at
+<https://help.github.com/en/articles/adding-a-new-ssh-key-to-your-github-account>.
+
+If you have any problems with SSH access to GitHub, read the troubleshooting
+instructions at <https://help.github.com/en/articles/troubleshooting-ssh>, or
+ask us on the [mailing list](https://groups.google.com/group/sympy).
+
+### Fork SymPy
+
+Create your own *fork* of the SymPy project on GitHub. If you have already
+done this before, you do not need to do it again.
+
+Go to the [SymPy GitHub repository](https://github.com/sympy/sympy) and click the **Fork** button.
+
+Now you have your own repository for the SymPy project. The address of the
+forked project will look something like
+`https://github.com/<your-github-username>/sympy`, where
+`<your-github-username>` is your GitHub username.
+
+## Get the SymPy Code
+
+It is recommended practice to create a fork of the SymPy project for your development purposes. Create your own fork of the SymPy project (if you have not yet). Go to the SymPy GitHub repository:
+
+```bash
+https://github.com/sympy/sympy
+```
+
+You will now have a fork at `https://github.com/<your-user-name>/sympy`.
+
+Then, on your machine browse to where you would like to store SymPy, and clone (download) the latest code from SymPy's original repository (about 77 MiB):
+
+```bash
+$ git clone https://github.com/sympy/sympy
+```
+
+Then assign your read-and-write repo to a remote called "github" (replace
+`<your-github-username>` with your GitHub username):
+
+```
+git remote add github git@github.com:<your-github-username>/sympy.git
+```
```

Comment:
The origin remote will already be configured by `git clone` or at least it would if `git clone` used the `git@` rather than `https` url. Lower down the github remote is not mentioned. I would remove this "github" remote part.

Reply:
This is the fork remote. The example below this definitely should say "github" and not "upstream" in the output though.

Reply:
Oh, wait this is suggesting an odd setup. Usually "origin" is the fork and "upstream" is the main repo (at least that's how I've always done it). I would say that the contributor should clone their fork and then add upstream as a remote.

I'm now wondering which of these gives a better workflow...

Reply:
I would try to stick with what is the common workflow and I think that using upstream for the main repo and origin for the fork is it e.g. see all answers in:
https://stackoverflow.com/questions/7244321/how-do-i-update-or-sync-a-forked-repository-on-github

Reply:
Our dev guide has always used the "clone the main repo/add your fork as a remote" workflow. There is also the opposite workflow which a lot of people use. I personally prefer the workflow that we've always suggested, because 

1) it matches what've already been recommending for a while
2) it's more uniform. With the "clone your fork" model, you have to fork literally every repo that you clone. And it doesn't even work for repos not on GitHub (which is admittedly rare these days. It also doesn't work for the situation where you push directly to a repo without forking it. Obviously these workflows aren't directly relevant for SymPy development but it's why I generally find this workflow simpler for general git usage.
3) Aesthetically it's a little more symmetric to just treat your own fork as "another upstream" (actually this guide calls the fork upstream `github` so that it can just use the same name everywhere, but it might be better to call it `yourusername` instead).

I've no idea which workflow is more common but I believe both are used, so I wouldn't really say there's an argument there.

Reply:
> I've no idea which workflow is more common

I'm pretty sure that the clone-your-fork model is more common and that there is more documentation and advice for that setup. Of course there can be advantages in doing it the other way but I think it is better to keep this in line with wider practice.

It seems particularly disjointed to recommend this setup without giving any instructions for how to use it to do anything useful. The most common thing that someone would want to do is keep their local repo up to date either for testing or for making new PRs. There are instructions elsewhere for doing that (e.g. the SO post I linked above) but those are targeted at the more common clone-your-fork setup.

Reply:
The [original guide](https://github.com/sympy/sympy/wiki/Development-workflow) uses it a lot but I guess those all got cut. Honestly, I'm starting to wonder if we should just cut this entire page.

Reply:
> I'm starting to wonder if we should just cut this entire page.

I think this page is good and has useful information although I would cut the "Configure Git Settings" part.

Otherwise the sections are

- Install Git
- Configure Your Name and Email in Git
- Setup GitHub
- Create a GitHub Account
- Setup SSH Keys
- Fork SymPy
- Get the SymPy Code
- Virtual Environment Setup

Those are all probably part of any "dev setup".

GitHub itself now seems to push access tokens rather than SSH keys. I don't know if there is any advantage there but otherwise everything above is essential information for contributors. Is this information duplicated elsewhere?

Reply:
Access tokens for pushing to git? I haven't seen that. How do you even set that up?

Reply:
See here:
https://docs.github.com/en/authentication
https://github.blog/2020-12-15-token-authentication-requirements-for-git-operations

Since GitHub stopped accepting username and password for https authentification their docs recommend personal access tokens by default. All of their docs suggest https URLs that can only be used with personal access tokens. They continue to allow ssh but I wouldn't be surprised if they decide against those at some point.

For now though I still use ssh and I suppose you do too and we can continue to recommend that in the sympy dev guide.

Reply:
I'm not sure that they are actually recommending to use a token instead of SSH for git. Tokens are used as a replacement for username and password, but SSH is also an option. Basically, instead of authenticating through git with your GitHub password, you now have to generate a special password token to use instead, which can have limited scope compared to your GitHub password.

I'm not a security expert, so maybe my read on this is wrong, but as far as I know, SSH is still more secure than using a token. It only requires uploading a public key to GitHub, so a breach of GitHub could not leak your private key. Personally I use [Secretive](https://github.com/maxgoedjen/secretive) which stores my SSH key in my Mac's secure enclave which makes it impossible for the private key to leak at all. None of this is possible with a token, which is basically just a special password to your account. The only advantage of tokens over SSH keys is that they can have restricted access and can automatically expire. But SSH keys already have some degree of restricted access. They can only be used to push to a repo, whereas tokens can potentially be used for any API access. You could in principle regularly rotate your SSH keys, although I'm not sure how many people do that. GitHub doesn't require it, so it must not be a huge threat.

Reply:
> tokens can potentially be used for any API access

You set the access when creating the token.

Reply:
Going back to the original point, though, as you say, these things are all part of any dev setup. So why do we need to document it specially in our docs? I think we could just as well find a decent guide someone else has made and point to it. That reduces the maintenance burden for us. And honestly, this guide has always been only decent at best, whereas I'm sure there are much better written guides out there. I expect this is the sort of thing the [Scientific Python](https://discuss.scientific-python.org/) project is interested in doing, if it hasn't been done already. 

The only SymPy specific thing in this entire guide is the bit about installing "mpmath" as a dependency. That can easily be moved as a small section in the other page. 

This git guide stuff was a lot more relevant when it was started a long time ago, because back then git was still quite new and even seasoned developers hadn't used it yet. But now, the only people who don't know how to use Git and GitHub are people who are new to open source development in general. This is a lot of people, to be sure, but this applies to basically all of open source. There are literally hundreds of guides like this one out there, or varying quality. It's shouldn't hard to find a decent one that we can recommend.

Reply:
There can be still some benefit in duplicating information from elsewhere if it makes the situation clearer for users.